v2/yarpctchannel/dialer.go

```diff
@@ -18,254 +18,187 @@
-// ensuring reuse of connections and other resources.
-//
-// Either the local service name (with the ServiceName option) or a user-owned
-// TChannel (with the WithChannel option) MUST be specified.
-func NewTransport(opts ...TransportOption) (*Transport, error) {
-	options := newTransportOptions()
+	// TODO hey, maybe this all belongs in Inbound.
+	// Maybe we create a channel separately in Dialer and Inbound since
+	// Hyperbahn is no longer a thing and we'll never receive requests from
+	// outbound connections.
 
-	for _, opt := range opts {
-		opt(&options)
+	// Create a TChannel
+	chopts := tchannel.ChannelOptions{
+		Tracer:              d.tracer,
+		OnPeerStatusChanged: d.onPeerStatusChanged,
 	}
-
-	if options.ch != nil {
-		return nil, fmt.Errorf("NewTransport does not accept WithChannel, use NewChannelTransport")
+	ch, err := tchannel.NewChannel(d.Caller, &chopts)
+	if err != nil {
+		// This should be unreachable since it indicates invalid channel
+		// options.
+		return err
 	}
+	d.ch = ch
 
-	return options.newTransport(), nil
-}
+	d.peers = make(map[string]*tchannelPeer)
 
-func (o transportOptions) newTransport() *Transport {
-	logger := o.logger
-	if logger == nil {
-		logger = zap.NewNop()
+	if d.ConnTimeout == 0 {
+		d.ConnTimeout = defaultConnTimeout
 	}
-	headerCase := canonicalizedHeaderCase
-	if o.originalHeaders {
-		headerCase = originalHeaderCase
-	}
-	return &Transport{
-		once:                lifecycle.NewOnce(),
-		name:                o.name,
-		addr:                o.addr,
-		listener:            o.listener,
-		connTimeout:         o.connTimeout,
-		connBackoffStrategy: o.connBackoffStrategy,
-		peers:               make(map[string]*tchannelPeer),
-		tracer:              o.tracer,
-		logger:              logger,
-		headerCase:          headerCase,
+	if d.ConnBackoff == nil {
+		d.ConnBackoff = yarpcbackoff.DefaultExponential
 	}
+
+	return nil
 }
 
-// ListenAddr exposes the listen address of the transport.
-func (t *Transport) ListenAddr() string {
-	return t.addr
+// Stop stops the TChannel transport. It starts rejecting incoming requests
+// and draining connections before closing them.
+// TODO In a future version of YARPC, Stop will block until the underlying channel
+// has closed completely.
+func (d *Dialer) Stop(ctx context.Context) error {
+	// Release all peers.
+	for _, peer := range d.peers {
+		peer.Release()
+	}
+	d.ch.Close()
+	d.connectorsGroup.Wait()
+	return nil
 }
 
 // RetainPeer adds a peer subscriber (typically a peer chooser) and causes the
 // transport to maintain persistent connections with that peer.
-func (t *Transport) RetainPeer(pid peer.Identifier, sub peer.Subscriber) (peer.Peer, error) {
-	t.lock.Lock()
-	defer t.lock.Unlock()
+//
+// RetainPeer must be called while the dialer is running.
+func (d *Dialer) RetainPeer(pid yarpc.Identifier, sub yarpc.Subscriber) (yarpc.Peer, error) {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	if d.ch == nil {
```

Comment:
This is fine for now but it would be preferable to explicitly track the
running state in a variable rather than checking if `internalState != nil`
because `internalState` can get more complex. Consider adding a TODO here.